pkg/operator/ceph/admission/validate.go

```diff
@@ -0,0 +1,104 @@
+	"strconv"
+)
+
+var (
+	cephclusterResource   = metav1.GroupVersionResource{Group: cephv1.CustomResourceGroup, Version: cephv1.Version, Resource: "cephclusters"}
+	rookClientSet         rookclient.Interface
+	universalDeserializer = serializer.NewCodecFactory(runtime.NewScheme()).UniversalDeserializer()
+)
+
+func validateUpdatedCephCluster(updatedCephCluster cephv1.CephCluster, found *cephv1.CephCluster) error {
+	if updatedCephCluster.Spec.DataDirHostPath != found.Spec.DataDirHostPath {
+		return errors.New(fmt.Sprintf("invalid update : DataDirHostPath change from %q to %q is not allowed", found.Spec.DataDirHostPath, updatedCephCluster.Spec.DataDirHostPath))
+	}
+
+	if updatedCephCluster.Spec.Network.HostNetwork != found.Spec.Network.HostNetwork {
+		return errors.New(fmt.Sprintf("invalid update : HostNetwork change from %q to %q is not allowed", strconv.FormatBool(found.Spec.Network.HostNetwork), strconv.FormatBool(updatedCephCluster.Spec.Network.HostNetwork)))
+	}
+	return nil
+}
+
+func ValidateCephResource(request *v1beta1.AdmissionRequest, context *clusterd.Context) error {
+	rookClientSet = context.RookClientset
+	// Checks if CephCluster is being updated and rejects the request if the dataDirHostPath is changed from initial value
+	if isCephClusterUpdate(request) {
+
+		// Fetch the existing ceph cluster object
+		found, err := fetchExistingCephCluster(request)
+		if err != nil {
+			return err
+		}
+		raw := request.Object.Raw
+		updatedCephCluster := cephv1.CephCluster{}
+		// Deserializing the incoming CephCluster object
+		if _, _, err := universalDeserializer.Decode(raw, nil, &updatedCephCluster); err != nil {
+			return errors.Wrap(err, "failed to deserialize pod object.")
+		}
+
+		err = validateUpdatedCephCluster(updatedCephCluster, found)
+		if err != nil {
+			return err
+		}
+	}
+
+	if isCephClusterCreate(request) {
+		raw := request.Object.Raw
+		createdCephCluster := cephv1.CephCluster{}
+		// Deserializing the incoming CephCluster object
+		if _, _, err := universalDeserializer.Decode(raw, nil, &createdCephCluster); err != nil {
+			return errors.Wrap(err, "failed to deserialize pod object.")
+		}
+
+		if createdCephCluster.Spec.External.Enable {
+			if createdCephCluster.Spec.Mon != (cephv1.MonSpec{}) || createdCephCluster.Spec.Dashboard != (cephv1.DashboardSpec{}) || createdCephCluster.Spec.Monitoring != (cephv1.MonitoringSpec{}) || createdCephCluster.Spec.DisruptionManagement != (cephv1.DisruptionManagementSpec{}) || len(createdCephCluster.Spec.Mgr.Modules) > 0 || len(createdCephCluster.Spec.Network.Provider) > 0 || len(createdCephCluster.Spec.Network.Selectors) > 0 {
+				return errors.New("invalid create : external mode enabled cannot have mon,dashboard,monitoring,network,disruptionManagement,storage fields in CR")
+			}
+
+		}
+	}
+	return nil
+}
+
+func isCephClusterCreate(request *v1beta1.AdmissionRequest) bool {
+	return request.Resource == cephclusterResource && request.Operation == "CREATE"
```

Comment:
I believe the create event is raised every time a client starts watching. For example, if we restart the operator we get a create event when the operator starts up. I assume the admission controller doesn't follow that pattern, but can you confirm if the admission controller also sees a create if it is restarted?

Reply:
Discussed this in the call